src/bundles/programmable_repl/programmable_repl.ts

```diff
@@ -0,0 +1,202 @@
+
+    So please DON'T add the "^" symbol before the version string for "ace-builds" and "react-ace" in package.json / dependencies.
+    And also DON'T edit the part in yarn.lock file related to the above two node modules.
+    It really cost me some time to fix the compatibility issue up with packages.json and yarn.lock .
+    If you modify the related part in these two files, your code may can't pass typechecking when building.
+
+    --Wang Zihan
+
+*/
+
+
+import { context } from 'js-slang/moduleHelpers';
+import { default_js_slang } from './functions';
+import { IOptions } from 'js-slang';
+
+
+
+export class ProgrammableRepl {
+  public evalFunction : Function;
+  public userCodeInEditor : String;
+  public outputStrings : any[];
+  private _editorInstance;
+  private _tabReactComponent : any;
+
+  public customizedEditorProps = {
+  	backgroundImageUrl: 'no-background-image',
+  	backgroundColorAlpha: 1,
+    fontSize: 17,
+  };
+
+  constructor() {
+    this.evalFunction = (placeholder) => this.easterEggFunc();
+    this.userCodeInEditor = this.getSavedEditorContent();
+    this.outputStrings = [];
+    this._editorInstance = null;// To be set when calling "SetEditorInstance" in the ProgrammableRepl Tab React Component render function.
+    developmentLog(this);
+  }
+
+  InvokeREPL_Internal(evalFunc : Function) {
+    this.evalFunction = evalFunc;
+  }
+
+  RunCode() {
+  	this.outputStrings = [];
+    let retVal : any;
+    try {
+      if (this.evalFunction === default_js_slang) {
+        retVal = this.evalFunction(this.userCodeInEditor, document.body);
+      } else {
+        retVal = this.evalFunction(this.userCodeInEditor);
+      }
+    } catch (exception : any) {
+    	this.OutputString(`Line ${exception.location.start.line.toString()}: ${exception.error.message}`, 'red');
+    	this.reRenderTab();
+      return;
+    }
+    if (retVal === undefined) {
+    	this.OutputString('Program exit with undefined return value.', 'cyan');
+    } else {
+      if (typeof (retVal) === 'string') {
+    		retVal = `\"${retVal}\"`;
+    	}
+    	this.OutputString(`Program exit with return value ${retVal}`, 'cyan');
+    }
+    this.reRenderTab();
+    developmentLog('RunCode finished');
+  }
+
+  UpdateUserCode(code) {
+    this.userCodeInEditor = code;
+  }
+
+  OutputString(str : String, textColor = 'white') {
+  	const checkResult = this.OutputStringSafeCheck(str);
+  	if (checkResult != 'safe') {
+  		str = `[Warning] Unsafe output string \"${checkResult}\" detacted. Please remove this in the output content and try again.`;
+  		textColor = 'red';
+  	}
+  	let tmp = {
+  		content: str,
+  		color: textColor,
+  	};
+  	this.outputStrings.push(tmp);
+  }
+
+  // Returns the forbidden word present in the string "str" if it contains at least one unsafe word. Returns "safe" if the string is considered to be safe to output directly into innerHTML.
+  private OutputStringSafeCheck(str) {

```

Comment:
We've discussed how this is XSS vulnerable, just putting this here for documentation.